File: Исходники/Глава 3. Часть 1/7. Быстрая сортировка/Python/QuickSort.py

def quick_sort(array: list):
    if len(array) < 2:
        return array
    pivot_index = len(array) // 2
    pivot = array[pivot_index]
    array.pop(pivot_index)
    big, small = list(), list()
    for i in array:
        if i > pivot:
            big.append(i)
        else:
            small.append(i)
    return quick_sort(small) + [pivot] + quick_sort(big)

File: Исходники/Глава 2. Часть 1/Разделяй и властвуй/2. Быстрая сортировка/C++/Олимпиадный/QuickSort.cpp
#include <vector>
#include <iostream>
#include <algorithm>


template <typename T>
int partition(std::vector<T> &array, int begin, int end) {
    T pivot = array[(begin + end) / 2];
    int i = begin;
    int j = end;
    while (true) {
        while (array[i] < pivot) {
            i++;
        }
        while (array[j] > pivot) {
            j--;
        }
        if (i >= j) {
            return j;
        }
        std::swap(array[i++], array[j--]);
    }
}


template <typename T>
void quick_sort(std::vector<T> &array, int begin, int end) {
    if (begin < end) {
        int pivot_index = partition(array, begin, end);
        quick_sort(array, begin, pivot_index);
        quick_sort(array, pivot_index + 1, end);
    }
    return;
}
    

